nexus/src/app/sagas/instance_update/mod.rs

```diff
@@ -0,0 +1,2514 @@
+//! represent rack-level resources allocated to the instance, and the instance's
+//! network configuration. When the instance's state changes, these resources
+//! owned by the instance may also need to be updated, such as changing the
+//! network configuration to point at an instance's new home after a successful
+//! migration, or deallocating virtual provisioning counters when an instance is
+//! destroyed. Naturally, these updates must also be performed reliably and
+//! inconsistent states must be avoided.
+//!
+//! Thus, we arrive here, at the instance-update saga.
+//!
+//! ## Theory of Operation
+//!
+//! In order to ensure that changes to the state of an instance are handled
+//! reliably, we require that all mutations of an instance` record are performed
+//! by a saga. The following sagas currently touch the `instance` record:
+//!
+//! - [`instance_start`]
+//! - [`instance_migrate`]
+//! - [`instance_delete`]
+//! - `instance_update` (this saga)
+//!
+//! For most of these sagas, the instance state machine itself guards against
+//! potential race conditions. By considering the valid and invalid flows
+//! through an instance's state machine, we arrive at some ground rules:
+//!
+//! - The `instance_migrate` and `instance_delete` sagas will only modify the
+//!   instance record if the instance *has* an active Propolis ID.
+//! - The `instance_start` and instance_delete` sagas will only modify the
+//!   instance record if the instance does *not* have an active VMM.
+//! - The presence of a migration ID prevents an `instance_migrate` saga from
+//!   succeeding until the current migration is resolved (either completes or
+//!   fails).
+//! - Only the `instance_start` saga can set the instance's *active* Propolis
+//!   ID, and it can only do this if there is currently no active Propolis.
+//! - Only the `instance_migrate` saga can set the instance's *target* Propolis
+//!   ID and migration ID, and it can only do that if these fields are unset.
+//! - Only the `instance_update` saga can unset a migration ID and target
+//!   Propolis ID, which it will do when handling an update from sled-agent that
+//!   indicates that a migration has succeeded or failed.
+//! - Only the `instance_update` saga can unset an instance's active Propolis
+//!   ID, which it will do when handling an update from sled-agent that
+//!   indicates that the VMM has been destroyed (peacefully or violently).
+//!
+//! For the most part, this state machine prevents race conditions where
+//! multiple sagas mutate the same fields in the instance record, because the
+//! states from which a particular transition may start limited. However, this
+//! is not the case for the `instance-update` saga, which may need to run any
+//! time a sled-agent publishes a new instance state. Therefore, this saga has
+//! the dubious honor of using the only distributed lock in Nexus (at the time
+//! of writing), the "instance updater lock".
```

Comment:
For better or worse I don't think this is the only such lock--Crucible volume replacement also has one (with similar constraints; it's usable because all the operations that work with it take place in sagas).

Reply:
Yeah, I was actually chatting with @jmpesp about potentially doing an RFD about distributed locking in sagas. This _was_ the first distributed lock in Nexus when I started working on it, but then James also added one --- I'll adjust this. :)